### DISCUSSION TITLE
question about the ili9341 stm32f4 spi driver

### DISCUSSION THREAD

**Discussion by franksterb92** [MaJerle/stm32f429#27] [Q&A] [answered]
in the tm_stm32f4_ili9341.c file the function putc in the for loop that draws the font data there is a if statement that shifts b by j and compares that with 0x8000 why is that value there why is it 0x8000 and not something else 

    for (i = 0; i < font->FontHeight; i++) {
        b = font->data[(c - 32) * font->FontHeight + i];
        for (j = 0; j < font->FontWidth; j++) {
            if ((b << j) & 0x8000) {
                TM_ILI9341_DrawPixel(ILI9341_x + j, (ILI9341_y + i), foreground);
            }
        }
    }

**Comment by MaJerle** [comment, marked as answer]
Fonts are described as a fixed size array, for instance variable `TM_Font7x10` has 10 `uint16_t` array values per character.
Each character line is described as bitfield where font is "active" and where we should draw background.

If we draw from left to right, we want to check always bits from left to right. This is done by shifting our `uint16_t` value by `j` and then check MSB bit. Mask for MSB bit is `0x8000`.

**Reply by franksterb92** [reply]
cool that makes since now thanks for writing all these drivers I've learned so much just looking through these

**Comment by hosein9279** [comment]
hi 
what software did you use to convert the font to code?

**Comment by EmbeddedLife615** [comment]
邮件已收到，我会尽快回复的！

**Comment by hosein9279** [comment]
Thank you for your response
